### DISCUSSION TITLE
Histogram serialization

### DISCUSSION THREAD

**Discussion by jonas-eschle** [hep-statistics-standard/hep-statistics-standard#26] [Standard]
We had some chats in the past on this (and also on the pyhf "hist-like-but-only-counts" style of histograms). At PyHEP.dev, we had some great discussions on this with Henry having already baked something together (for the boost-histogram library) here: https://github.com/scikit-hep/uhi/pull/105.

Two things:
1. what do you think of it in general? Technically, we don't need to follow the same definition but converging with BH on the same definition should definitely be the way to go (if somehow possible). They surely should also for for RooFit/ROOT, @cburgard 
2. Can we formulate pyhf histograms "consistently" with it, i.e. as a "subset" or vice-versa? (just to keep as much consistency as possible). Maybe @lukasheinrich has some points on it?

**Comment by lmoureaux** [comment]
What justifies the need for a new storage format? I know of three well established formats: ROOT, YODA, and HepData YAML. Both YODA (in Rivet) and HepData (on HepData) are already used for data preservation. Sounds like standard proliferation...

**Reply by jonas-eschle** [reply]
Good point! My thoughts:
- ROOT, AFAIK is only binary but we should ideally communicate with them.
- YODA seems purely text, not JSON-like and therefore isn't really compatible
- hepdata seems to only store the counts, in the HistFactory/pyhf-like manner. This aren't "real histograms" but just the integrated values (that suffices for HistFactory-like statistical inferenc but doesn't to store a whole boost-histogram or do any statistical inference that has other, analytic PDF/unbinned datasets). However, surely worth to work closer with HEPData, they're probably also interested?

**Reply by jpivarski** [reply]
None of these (ROOT, YODA, HepData) can represent an n-dimensional boost-histogram ([C++](https://www.boost.org/doc/libs/1_82_0/libs/histogram/doc/html/index.html), [Python](https://boost-histogram.readthedocs.io/)). Uproot converts [ROOT.TH* into boost-histograms](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TH1.TH1.html#to-boost) and [boost-histograms into ROOT.TH*](https://uproot.readthedocs.io/en/latest/basic.html#writing-objects-to-a-file), but this is a very lossy operation. Histograms beyond 3 dimensions (beyond the limit that can be visualized) are very useful for statistical fits.[^1]

Since boost-histograms are now fairly widely used (in the Coffea framework, for instance), not having a way to write them to disk and read them back again is a pain point.

It would be a problem trying to fit these serialized histograms into a specification that is purely JSON (as I understand HS3 to be). That's a different point. Maybe HS3 can reference histogram objects outside of itself? The main thing that I want to argue here is that a serialization format for boost-histograms is well motivated.

[^1]: I wish I could find it now, but I came across a physicist's analysis code that naturally needed 5 dimensions, but it was cobbled together with TDirectories of TH3s, overly complicating what they were trying to do.

**Reply by lmoureaux** [reply]
Thanks! I see the point of a new format then. I think HS3 could use whatever works for HistFactory (let's call this "the HS3 format") and if a new format gets developed, switch to it at some point. Would definitely try to include the Rivet/YODA people if developing a new standard, from what I understand from [Section 3.2.2 of the Rivet 3 paper](https://arxiv.org/pdf/1912.05451.pdf) histogramming isn't so trivial for MC generators with counter-events (e.g. FxFx).

**Comment by cburgard** [comment]
Just to add to the discussion:

RooFit itself can deal with arbitrary-dimension histograms in principle, they are just extremely cumbersome to construct as none of the standard constructors based on ROOT TH* can be used.

The current HS3 draft reflects this possibility by allowing arbitrary dimensionality, like so:

```
"data":[
  {
    "name":"data2",
    "type":"binned",
    "contents":[
      9.0,
      18.4
    ],
    "axes":[
      {
        "name":"variable1",
        "nbins":2,
        "min":1,
        "max":3
      },
      ...
    ]
  },
  ...
]
```

As you can see, the "axes" field is a list, so any number of dimensions can be included. The only non-trivial aspect about this is that the order of the axes matter, as they motivate the order in which the one-dimensional array "contents" should be mapped to the multi-dimensional space mapped by the axes.

This is a relatively powerful mechanism that is straight copied from the way RooFit does things, so we have a 1-1 mapping there, which is convenient.

However, as @lukasheinrich pointed out in the past, this format is not omnipotent in that it cannot cover cases where the bin boundaries in one dimension depend on another dimension - think of a 2d histogram where the top right corner quadrant is merged into one big bin, or @lukasheinrich s example of a hexagonal honeycomb binning.

So if anyone knows of a nice way to write down a histogram such that also these cases are covered, please elaborate :)

**Comment by cburgard** [comment]
Just to follow up on this: The current suggestion for irregular binned axes is like this:

```
    "axes":[
      {
        "name":"variable1",
        "bounds":[0,50,100,200,1000]
      },

```

@Cornelius-G has raised several important points here, namely:

- should we use "edges" instead of "bounds" to be consistent with typical framework namings?
- should we take any measures to define whether the edges are left-edges or right-edges of the histogram?

Especially the last case is, at least in my mind, a bit controversial, as the question of whether the bins are `[a,b[` or `]a,b]` is relevant at the time of filling the histograms, but should not matter at the time of modelling them, as the bin edges themselves are null sets irrelevant to the pdf.

One could make the point that someone might want to fill the histogram with more entries after serializing and deserializing it, and it should thus be defined, but to me, this would be more something to go in the `misc` section rather than an inherent property of the dataset.

What do other people think?

**Reply by jpivarski** [reply]
"edges" is a good name. In Scikit-HEP, we have [defined protocols for these things](https://uhi.readthedocs.io/en/latest/plotting.html) so that libraries that produce histograms and libraries that consume histograms all agree on the naming and other conventions. "edges" is the name used in that protocol.

Also, wouldn't the edges be both left and right—i.e. if a histogram has $n$ bins, you'd have $n+1$ edges by the fencepost principle?

I can see where you're coming from about whether the left or right edge is inclusive while filling the histogram. It might only matter while filling, but the filling algorithm is important when interpreting the filled histogram. All histogram libraries that I'm familiar with consider the left edge to be included and the right edge to be not included when filling. So if one bin covers the range from 4.0 to 5.0 and another covers the range from 5.0 to 6.0, filling it with 5.0 would put it in the second bin only.

Here's that in ROOT:

```python
>>> import ROOT
>>> th1d = ROOT.TH1D("th1d", "", 10, 0, 10)
>>> th1d.Fill(5.0)
6
>>> [th1d.GetBinContent(i + 1) for i in range(10)]
[0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0]
```

and here's that in boost-histogram/hist:

```python
>>> import hist
>>> h = hist.Hist.new.Reg(10, 0, 10).Double()
>>> h.fill(5.0)
Hist(Regular(10, 0, 10, label='Axis 0'), storage=Double()) # Sum: 1.0
>>> h.values()
array([0., 0., 0., 0., 0., 1., 0., 0., 0., 0.])
```

They both have inclusive-left, exclusive-right. If anything had the opposite, the bin contents would become

```python
[0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]
```

and I've never seen a histogram library do that.

There's a similar, subtle question about underflow and overflow bins: technically, the underflow bin covers the range from $-\infty$ (inclusive) to the first left edge (exclusive) and the overflow bin covers the range from the last right edge to $\infty$. If we fill with $-\infty$ or $\infty$, do the underflow and overflow bins increment? There's no question that the underflow bin should increment, since $-\infty$ is the left edge of its range and left edges are always inclusive. But for the overflow bin, if $\infty$ is its right edge and we fill it with $\infty$, that would usually be exclusive. However, both ROOT and boost-histogram/hist consider the overflow bin to be exceptional and any value greater than or equal to the last right edge is included.

```python
>>> th1d.Fill(-np.inf)
-1
>>> th1d.Fill(np.inf)
-1
>>> th1d.GetBinContent(0)    # underflow bin
1.0
>>> th1d.GetBinContent(11)   # overflow bin
1.0
```

and

```python
>>> h.fill(-np.inf)
Hist(Regular(10, 0, 10, label='Axis 0'), storage=Double()) # Sum: 1.0 (2.0 with flow)
>>> h.fill(np.inf)
Hist(Regular(10, 0, 10, label='Axis 0'), storage=Double()) # Sum: 1.0 (3.0 with flow)
>>> h.values(flow=True)      # show underflow and overflow as first and last elements
array([1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1.])
```

Interestingly, they also both consider `NaN` to belong to the overflow bin, so it's a bit of a catch-all for anything that wouldn't go the underflow or any other bin:

```python
>>> th1d.Fill(np.nan)
-1
>>> th1d.GetBinContent(0)
1.0
>>> th1d.GetBinContent(11)
2.0
```

and

```python
>>> h.fill(np.nan)
Hist(Regular(10, 0, 10, label='Axis 0'), storage=Double()) # Sum: 1.0 (4.0 with flow)
>>> h.values(flow=True)
array([1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 2.])
```

**Reply by lmoureaux** [reply]
About underflow and overflow, note that JSON doesn't natively support infinities so special support would be needed to define the underflow and overflow bin(s).

**Reply by cburgard** [reply]
These are all interesting and important points, thanks for raising them. In the light of this, I would be strongly in favour of using "edges" and sticking with the fencepost principle as discussed, that is:
- the left bin edge is always inclusive
- the right bin edge is always exclusive, except for the overflow bin, which is by filled whenever any value occurs that does not fit into any other bin, including NaN
As far as JSON support for infinities is concerned, I would be tempted to go with "inf" (as a string) for clarity, but there might be other, better options.
